lib/ecto/repo.ex

```diff
@@ -2200,6 +2214,81 @@ defmodule Ecto.Repo do
               | {:error, any}
               | Ecto.Multi.failure()
 
+  @doc """
+  Runs the given function inside a transaction.
+
+  The return value is the same as of the given `fun` which must be
+  `{:ok, result}` or `{:error, reason}`.
+
+  If this function returns `{:ok, result}`, it means the transaction
+  was successfully committed. On the other hand, if it returns `{:error, reason}`,
+  it means the transaction was rolled back.
+
+  If an Elixir exception occurs the transaction will be rolled back
+  and the exception will bubble up from the transaction function.
+  If no exception occurs, the transaction is committed when the
+  function returns. A transaction can be explicitly rolled back
+  by calling `c:rollback/1`, this will immediately leave the function
+  and return the value given to `rollback` as `{:error, value}`.
+
+  See `c:transaction/2` for more information on transactions.
+
+  ## Options
+
+  See the ["Shared options"](#module-shared-options) section at the module
+  documentation for more options.
+
+  ## Examples
+
+  This function is commonly used with `with/1`:
+
+      Repo.transaction_with(fn ->
+        with {:ok, alice} <- Repo.insert(alice_changeset),
+             {:ok, bob} <- Repo.insert(bob_changeset) do
+          {:ok, [alice, bob]}
+        end
+      end)
+
+  If the transaction was successful, `{:ok, result}` is returned:
+
+      iex> Repo.transaction_with(fn ->
+      ...>   Repo.insert(changeset)
+      ...> end)
+      {:ok, %User{}}
+
+  If the transaction failed, `{:error, reason}` is returned:
+
+      iex> Repo.transaction_with(fn ->
+      ...>   Repo.insert(changeset)
+      ...> end)
+      {:error, #Ecto.Changeset<...>}
+
+  Transaction can be aborted by returning `{:error, reason}`, calling `c:rollback/1`,
+  or raising from the given `fun`:
+
+      iex> Repo.transaction_with(fn ->
+      ...>   Repo.insert!(%User{}) # will be rolled back
+      ...>   {:error, :oops}
+      ...> end)
+      {:error, :oops}
+
+      iex> Repo.transaction_with(fn ->
+      ...>   Repo.insert!(%User{}) # will be rolled back
+      ...>   Repo.rollback(:oops)
+      ...> end)
+      {:error, :oops}
+
+      iex> Repo.transaction_with(fn ->
+      ...>   Repo.insert!(%User{}) # will be rolled back
+      ...>   raise "oops"
+      ...> end)
+      ** (RuntimeError) oops
+  """
+  @doc group: "Transaction API"
+  @callback transaction_with(fun, opts :: Keyword.t()) ::
```

Comment:
I am wondering if changing from `fun` to (-> {:ok, term()} | {:error, term()})` is a better hint.

Reply:
Good call, done in 9a9f133e.